### DISCUSSION TITLE
Exergy Analysis for heatpump

### DISCUSSION THREAD

**Discussion by Adefunmike** [oemof/tespy#568] [Q&A] [answered]
So i was trying to perform exergy analysis for the connection below and i think i have correctly set all the parameters but i am getting error 
You have not provided enough parameters: 8 required, 4 supplied. Aborting calculation!
Can you please help
![Connection setup](https://github.com/user-attachments/assets/4671e538-1f2a-4ac1-b9cd-674b620063f4)

```python
from tespy.networks import Network
from tespy.components import Compressor, SimpleHeatExchanger, Valve, CycleCloser
from tespy.connections import Connection
from tespy.tools import ExergyAnalysis, document_model

# Define the working fluid and system boundaries
working_fluid = ['R134a']
ambient_temp = 298.15  # K (25°C)
ambient_pressure = 1.013  # bar

# Initialize the network
nw = Network(fluids=working_fluid, p_unit='bar', T_unit='C', h_unit='kJ / kg', m_unit='kg / s')

# Components: compressor, condenser, evaporator, expansion valve, and cycle closer
comp = Compressor('compressor')
cond = SimpleHeatExchanger('condenser')
evap = SimpleHeatExchanger('evaporator')
valve = Valve('expansion valve')
cycle_closer = CycleCloser('cycle closer')

# Set efficiency for compressor only
comp.set_attr(eta_s=0.85)

# Define Connections with balanced parameters

# 1. Cycle closer to compressor (define mass flow and fluid composition)
cycle_to_comp = Connection(cycle_closer, 'out1', comp, 'in1')
cycle_to_comp.set_attr(m=0.1, fluid={'R134a': 1.0})
nw.add_conns(cycle_to_comp)

# 2. Compressor to condenser (define outlet pressure only)
comp_out = Connection(comp, 'out1', cond, 'in1')
comp_out.set_attr(p=8.0)
nw.add_conns(comp_out)

# 3. Condenser to expansion valve (define outlet temperature only)
cond_out = Connection(cond, 'out1', valve, 'in1')
cond_out.set_attr(T=40)
nw.add_conns(cond_out)

# 4. Expansion valve to evaporator (define outlet pressure only)
valve_out = Connection(valve, 'out1', evap, 'in1')
valve_out.set_attr(p=1.8)
nw.add_conns(valve_out)

# 5. Evaporator to cycle closer (define outlet temperature only)
evap_out = Connection(evap, 'out1', cycle_closer, 'in1')
evap_out.set_attr(T=5)
nw.add_conns(evap_out)

# Solve the network
nw.solve('design')

# Capture compressor power and condenser heat values for exergy analysis
power_compressor = comp.P.val_SI if comp.P.val_SI is not None else 0
cond_heat = abs(cond.Q.val_SI) if cond.Q.val_SI is not None else 0

# Perform exergy analysis if both compressor power and condenser heat are available
if power_compressor > 0 and cond_heat > 0:
    ex = ExergyAnalysis(network=nw, E_F=[power_compressor], E_P=[cond_heat])
    ex.analyse()

    # Document the model and print exergy results
    document_model(nw, filename='air_cooled_heat_pump')
    print("Exergy Analysis - Connection Data:")
    print(ex.connection_data)
    print("Exergy Analysis - Component Data:")
    print(ex.component_data)
else:
    print("Exergy analysis could not be performed due to missing compressor power or condenser heat.")
```

**Comment by fwitte** [comment, marked as answer]
Hi @Adefunmike,

you are missing saturation state information after the condenser `x=0` and the evaporator `x=1` as well as pressure drop in those components. For that reason I would also not specify the pressure after compression and expansion, but let that be a result from the heat exchangers' pressure drops. Finally, the compressor outlet state is not determined, it misses a temperature or the compressor misses an efficiency value.

Best

**Reply by Adefunmike** [reply]
I made the adjustments blelow

from tespy.networks import Network
from tespy.components import Compressor, SimpleHeatExchanger, Valve, CycleCloser
from tespy.connections import Connection
from tespy.tools import ExergyAnalysis, document_model

# Define the working fluid and initialize the network
working_fluid = ['R134a']
nw = Network(fluids=working_fluid, p_unit='bar', T_unit='C', h_unit='kJ / kg', m_unit='kg / s')

# Define components with minimal but necessary attributes
comp = Compressor('compressor')
cond = SimpleHeatExchanger('condenser')
evap = SimpleHeatExchanger('evaporator')
valve = Valve('expansion valve')
cycle_closer = CycleCloser('cycle closer')

# Set key attributes for components
comp.set_attr(eta_s=0.85)  # Efficiency for compressor
cond.set_attr(pr=0.98, Q=-100)  # Pressure ratio for condenser with estimated heat load
evap.set_attr(pr=0.98)  # Pressure ratio for evaporator

# Define connections with essential properties

# 1. Cycle closer to compressor: Set mass flow rate and fluid composition
cycle_to_comp = Connection(cycle_closer, 'out1', comp, 'in1')
cycle_to_comp.set_attr(m=0.1, fluid={'R134a': 1.0})
nw.add_conns(cycle_to_comp)

# 2. Compressor to condenser: Specify the outlet pressure
comp_out = Connection(comp, 'out1', cond, 'in1')
comp_out.set_attr(p=8)  # Target pressure at condenser inlet
nw.add_conns(comp_out)

# 3. Condenser to valve: Define outlet state as saturated liquid
cond_out = Connection(cond, 'out1', valve, 'in1')
cond_out.set_attr(x=0.0)  # Saturated liquid state
nw.add_conns(cond_out)

# 4. Valve to evaporator: Define pressure after expansion
valve_out = Connection(valve, 'out1', evap, 'in1')
valve_out.set_attr(p=1.8)  # Pressure at evaporator inlet
nw.add_conns(valve_out)

# 5. Evaporator to cycle closer: Define outlet as saturated vapor
evap_out = Connection(evap, 'out1', cycle_closer, 'in1')
evap_out.set_attr(x=1.0)  # Saturated vapor after evaporator
nw.add_conns(evap_out)

# Solve the network
nw.solve('design')

# Perform exergy analysis if compressor power and condenser heat are valid
if comp.P.val_SI is not None and cond.Q.val_SI is not None:
    ex = ExergyAnalysis(network=nw, E_F=[comp.P.val_SI], E_P=[abs(cond.Q.val_SI)])
    ex.analyse()

    # Document the model and print exergy results
    document_model(nw, filename='air_cooled_heat_pump')
    print("Exergy Analysis - Connection Data:")
    print(ex.connection_data)
    print("Exergy Analysis - Component Data:")
    print(ex.component_data)
else:
    print("Exergy analysis could not be performed due to missing compressor power or condenser heat.")
 But i still got error
  You have provided too many parameters: 8 required, 9 supplied. Aborting calculation!

**Comment by Adefunmike** [comment]
Thanks for the help so far
I really appreciate it. Please i need more favor!
I tried calculating the exergy using the tutorial but was getting error. Below is my code and the error I got
from tespy.tools import ExergyAnalysis
from tespy.connections import Bus

# Define ambient conditions
pamb = 1e5  # Ambient pressure in Pa
Tamb = [25 + 273.15]  # Ambient temperature in K

# Define exergy buses
E_F = Bus("Fuel Exergy")
E_P = Bus("Product Exergy")

# Add components to buses
E_F.add_comps({'comp': compressor})  # Compressor work as fuel
E_P.add_comps({'comp': condenser})  # Heat delivered by condenser as product

# Perform exergy analysis
ex = ExergyAnalysis(network=nw, E_F=[E_F], E_P=[E_P])
ex.analyse(pamb=pamb, Tamb=Tamb)

# Print results
ex.print_results()
TypeError: can only concatenate list (not "float") to list
